opentelemetry-sdk/src/opentelemetry/sdk/_metrics/metric_reader_storage.py

```diff
@@ -12,23 +12,105 @@
-        pass
+        self._lock = RLock()
+        self._sdk_config = sdk_config
+        self._view_storages: Dict[Instrument, List[ViewStorage]] = {}
+
+    def _get_or_init_view_storage(
+        self, instrument: Instrument
+    ) -> List["ViewStorage"]:
+        # Optimistically get the relevant views for the given instrument. Once set for a given
+        # instrument, the mapping will never change
+        if instrument in self._view_storages:
+            return self._view_storages[instrument]
+
+        with self._lock:
+            # double check if it was set before we held the lock
+            if instrument in self._view_storages:
+                return self._view_storages[instrument]
+
+            # not present, hold the lock and add a new mapping
+            view_storages = []
+            for view in self._sdk_config.views:
+                if view.match(instrument):
+                    # Note: if a view matches multiple instruments, this will create a separate
+                    # ViewStorage per instrument. If the user's View configuration includes a
+                    # name, this will cause multiple conflicting output streams.
+                    view_storages.append(
+                        ViewStorage(
+                            view=view,
+                            instrument=instrument,
+                            sdk_config=self._sdk_config,
+                        )
+                    )
+
+            # if no view targeted the instrument, use the default
+            if not view_storages:
+                view_storages.append(
+                    ViewStorage(
+                        view=default_view(instrument),
+                        instrument=instrument,
+                        sdk_config=self._sdk_config,
+                    )
+                )
+            self._view_storages[instrument] = view_storages
+            return view_storages
 
     def consume_measurement(self, measurement: Measurement) -> None:
-        pass
+        storages = self._get_or_init_view_storage(measurement.instrument)
+        for storage in storages:
+            storage.consume_measurement(measurement)
 
     def collect(self, temporality: AggregationTemporality) -> Iterable[Metric]:
-        pass
+        # use a list instead of yielding to prevent a slow reader from holding SDK locks
+        metrics: List[Metric] = []
+
+        # While holding the lock, new ViewStorage can't be added from another thread (so we are
+        # sure we collect all existing view). However, instruments can still send measurements
+        # that will make it into the individual aggregations; collection will acquire those
+        # locks iteratively to keep locking as fine-grained as possible. One side effect is
+        # that end times can be slightly skewed among the metric streams produced by the SDK,
+        # but we still align the output timestamps for a single instrument.
+        with self._lock:
```

Comment:
Aren't we missing here a `for` loop that will go through all the [asynchronous instruments](https://github.com/open-telemetry/opentelemetry-python/pull/2301/files#diff-693118fd4a59966d0e9d2499bb9704ce0af972e3390e712b3a7488e0848009bbR119-R123) first?

Reply:
Moved async instrument collection to the MetricReaderStorage and updated async instruments to be tracked in `SdkConfiguration` object. This will have to be changed in the future when we implement https://github.com/open-telemetry/opentelemetry-specification/pull/2317.

Reply:
Hm, ok, but, isn't the `SdkConfiguration` getting a bit "out of control"? It started as something that gathered (somewhat unrelated) objects, but now it also can receive and track asynchronous instruments as well :shrug: 

Besides its name making the reader think it is actually the configuration of the SDK (which is not), I am getting a feeling that this class is turning a little into a something where we can just put stuff inside and move it around, and I don't feel that this is necessarily the most convenient approach, in my opinion, of course.

Reply:
In today Python SIG we discussed where to put the attributes of `SdkConfiguration`. I still think there is room for discussion there, but I think we can leave `SdkConfiguration` as it is right now and address that later and move this PR forward if we make a few things private:
1. Make `MetricReaderStorage` private
2. Make `SdkConfiguration` private
3. Make `SynchronousMeasurementProcessor` private

I think all of these classes are either components of the "internal" part of the design or they are an argument passed to their constructor so it should be fine to do this. :+1:

Reply:
> Make MetricReaderStorage private
> Make SdkConfiguration private
> Make SynchronousMeasurementProcessor private

these are all intended to be private, going to address in #2389.

Reply:
All right, resolving 👍